### ISSUE TITLE
native amp consumes 10x gpu memory 

### ISSUE THREAD

**Issue by stas00** [pytorch/pytorch#48049]
## 🐛 Bug

observation: pytorch native amp consumes 10x memory as compared to fp16/apex or no `--fp16` usage.

The issue has been originally discovered and tracked at https://github.com/huggingface/transformers/issues/8403 and on slack, but now we know for sure that this is a pytorch issue, so let's concentrate the discussion here.

## To Reproduce

What's following is the minimal script that was reduced from a rather complex logic of `finetune.py` in `transformers`. It originates in a complex stack of `pytorch-lightning` feeding `transformers` generate/beam_search logic finally feeding to real pytorch logic in `BartForConditionalGeneration`. If you want to see the full blow up and how to set it up, it's described [here](https://github.com/huggingface/transformers/issues/8403).

I found at least part of the culprit or trigger of the leak - it's `@torch.no_grad()` used for `generate` https://github.com/huggingface/transformers/blob/eb3bd73ce35bfef56eeb722d697f2d39a06a8f8d/src/transformers/generation_utils.py#L281-L282

Here is a short script that reproduces the leakage. It removes all the generate/search logic and feeds the same random `input_ids` to `BartForConditionalGeneration` pre-trained model.

Please first run:
```
pip install ipyexperiments 
```
to get the memory tracing, but feel free to disable it if for some reason it's not working for you and use some alternative like `nvidia-smi` dumping. (but it should work)

```python
#!/usr/bin/env python

import os
import sys
import torch
os.environ["USE_TF"] = "0"
sys.path.insert(1, "src")

# !pip install ipyexperiments 
from ipyexperiments.utils.mem import gpu_mem_get_used_mbs, gpu_mem_get_used_no_cache_mbs

from transformers import BartForConditionalGeneration

device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')

model = BartForConditionalGeneration.from_pretrained('sshleifer/student_cnn_12_6').to(device)
model.eval()

vocab_size = 50264 # model.config.vocab_size
length = 10

AUTOCAST = False if "-f" in sys.argv else True
print(f"autocast: {AUTOCAST}")

class MemReport():
    def __init__(self, gc_collect=True):
        self.get_mem = gpu_mem_get_used_no_cache_mbs if gc_collect else gpu_mem_get_used_mbs
        self.cur = self.get_mem()
    def delta(self, id):
        peak = torch.cuda.memory_stats()["allocated_bytes.all.peak"]
        print(f"{id}: {gpu_mem_get_used_mbs()-self.cur}MB (peak {peak>>20}MB)")
        self.cur = self.get_mem()
        
mr = MemReport(gc_collect=False)

### reproducible code starts here ###

@torch.no_grad()
def logic():
    input_ids = torch.randint(vocab_size, (1,length)).to(device)
    mr.delta(0)
    for i in range(1,10):
        outputs = model(input_ids)
        mr.delta(i)

if AUTOCAST:
    with torch.cuda.amp.autocast():
        logic()
else:
    logic()
```

So if I run it with `-f` which disables `autocast`, I get:

```
./reproduce.py -f
autocast: False
0: 0MB (peak 1165MB)
1: 12MB (peak 1167MB)
2: 0MB (peak 1169MB)
3: 0MB (peak 1169MB)
4: 0MB (peak 1169MB)
5: 0MB (peak 1169MB)
6: 0MB (peak 1169MB)
7: 0MB (peak 1169MB)
8: 0MB (peak 1169MB)
9: 0MB (peak 1169MB)
```
no leak.

If however I remove `-f` and `autocast` gets enabled, we get:
```
./reproduce.py
autocast: True
0: 0MB (peak 1165MB)
1: 592MB (peak 1744MB)
2: 580MB (peak 2324MB)
3: 580MB (peak 2902MB)
4: 580MB (peak 3480MB)
5: 580MB (peak 4058MB)
6: 580MB (peak 4636MB)
7: 580MB (peak 5214MB)
8: 580MB (peak 5793MB)
9: 580MB (peak 6371MB)
```
the memory logger prints the delta for each `forward` call in the loop and the peak memory.

You can see that we are leaking 600Mb per forward call here.

If I comment out `@torch.no_grad()`, the total memory usage doubles but there is no leak:

```
autocast: True
0: 0MB (peak 1165MB)
1: 602MB (peak 1754MB)
2: 590MB (peak 2343MB)
3: 0MB (peak 2343MB)
4: 0MB (peak 2343MB)
5: 0MB (peak 2343MB)
6: 0MB (peak 2343MB)
7: 0MB (peak 2343MB)
8: 0MB (peak 2343MB)
9: 0MB (peak 2343MB)
```

I was using pycharm to debug this and to write a small script and boy it got me so delayed as it leaks gpu ram on its own, since it has to save all those variables on cuda, but I wasn't aware of it. Well, now I know not to do that. Luckily I had https://github.com/stas00/ipyexperiments handy to give me easy memory tracing.

Note I'm importing two gpu mem tracking functions - one of them clears cuda cache - but here it appears it's better not use that version. 

## Expected behavior

Shouldn't consume 10x gpu memory (apex does not).

## Environment

```
Collecting environment information...
PyTorch version: 1.8.0.dev20201115+cu110
Is debug build: False
CUDA used to build PyTorch: 11.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: GeForce RTX 3090
GPU 1: GeForce GTX 1070 Ti

Nvidia driver version: 455.32.00
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.4
[pip3] pytorch-lightning==1.0.4
[pip3] pytorch-memlab==0.2.2
[pip3] torch==1.8.0.dev20201115+cu110
[pip3] torchtext==0.6.0
[pip3] torchvision==0.9.0.dev20201115+cu110
[conda] numpy                     1.19.4                   pypi_0    pypi
[conda] pytorch-lightning         1.0.4                     dev_0    <develop>
[conda] pytorch-memlab            0.2.2                    pypi_0    pypi
[conda] torch                     1.8.0.dev20201115+cu110          pypi_0    pypi
[conda] torchtext                 0.6.0                    pypi_0    pypi
[conda] torchvision               0.9.0.dev20201115+cu110          pypi_0    pypi
```

@mcarilli, @ptrblck 

cc @ezyang @gchanan @zou3519 @bdhirsh @mcarilli

**Comment by ezyang** [comment]
I read through the downstream discussion. I agree that there are (multiple) legitimate bugs; but furthermore, I think we were wrong to close earlier reports that AMP leaks memory when you use AMP outside of your training loop. This should work. @mcarilli, I'll try to explain why below.

On its face, it would make sense that a cache that only gets cleared when you exit AMP would leak memory over time. But AMP only caches casts on parameters: you should be using the same parameters in each loop of the training loop (and updating them in place). What is there to leak? In fact, there should be no leak, and instead we have a correctness bug where we will improperly use casts even when the original parameters have been updated. We ought to fix that.

But let's suppose that, somehow, you're generating fresh new parameters every iteration of the loop. Those parameters are ostensibly going dead when you go around the loop, and so if we implemented a weak map correctly, they should still get pruned anyway. Well I looked over the code and there are still some problems:

* We don't seem to ever traverse the map to get rid of dead entries, where the key is dead. This is a big deal because even though the source parameter's data gets deallocated when it becomes dead, the *cached cast output* doesn't get deallocated unless you remove it from the map. So the weak pointer is only solving half the problem: we prevented the source parameter from getting leaked, but we're still leaking the cached output.
* I stared at the code some more and I got worried about another possible memory leak: the cached output cast has a `grad_fn` that will refer to the parameter via a grad accumulator. I now don't remember if this is a weak reference or a strong ref; if it's strong, that's a reference cycle and would also prevent the weak pointer key from ever getting collected either.

We probably didn't notice any of this because it's mostly OK to just leak everything until the end of one step in the training loop... except in situations like this.

PS. I'm kind of sad we don't live in a GC'ed language, as a lot of this weak pointer/map business gets a lot simpler if you can assume there's some GC process that will clean up after you.

btw, cc @swolchok, this is an example of weak pointers actually being important for tensors

**Comment by agolynski** [comment]
@mcarilli

**Comment by mcarilli** [comment]
I was still confused because https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/autocast_mode.cpp#L71 should ensure the only things that get cached are leaves (model params) that require grad.  These don't move between iterations of the loop under `no_grad`.  @ptrblck investigated yesterday in depth yesterday and we realized there are 3 ingredients that cause the cache blowup we didn't discuss so far:
1. All tensors created in a `no_grad` context are considered leaves, because they don't have autograd history.
2. Linear layers don't send their weights directly to the matmul call, rather, they send [`weight.t()`](https://github.com/pytorch/pytorch/blob/9b526546202cf53c90ed3a438c16260386475680/torch/nn/functional.py#L1665-L1667), which is a new Tensor (with a new TensorImpl) each time.  (further unfortunate consequence:  this means without `no_grad`, the cache was doing nothing for linear layers, because it never saw incoming `weight.t()`s as leaves.)
3. View ops (like `t()`) in `no_grad` context _do_ propagate requires_gradness from their source tensor.

With 1, 2, and 3 together, when `forward` with many linear layers is called in a loop under `no_grad`, in every matmul in every iteration `cached_cast` sees incoming weights as leaves that require grad and have a new TensorImpl, so it creates a new cache entry with a new fp16 copy every time.

The plot thickens, though.  3. is an _inconsistent behavior_ between view ops and out-of-place ops in `no_grad` contexts.  View ops propagate requires_gradness of the source, ordinary out-of-place ops don't:
```python
with torch.no_grad():
    print(a.requires_grad) # True
    b = a.clone()
    print(b.requires_grad) # False
    b = a.t()
    print(b.requires_grad) # True
    b = a.view(-1)
    print(b.requires_grad) # True
```
This may be due to some short-circuiting in view ops, I haven't looked in detail.
@albanD , is the above behavior expected, or a bug?  Seems like a bug to me: since view ops do create distinct tensors, they should have the same source-tensor-attribute-propagation policy as out-of-place ops.  But maybe the different policy is a deliberate choice for some reason I don't get.

One option:  If we changed view ops to _not_ propagate requires_gradness under `no_grad` (consistent with the behavior of out-of-place ops) it would fix the cache blowup, because the incoming `weight.t()`s would have `requires_grad=False`, therefore `cached_cast` would not try to cache their casts.  It wouldn't be ideal because it would defeat the purpose of caching, but it would be no worse than running without `no_grad` (see "2....further unfortunate consequence")

**Comment by t-vi** [comment]
The special handling of views in `no_grad` is required, this is part of `DifferentiableViewMeta` in `torch/csrc/autograd/variable.h`:

https://github.com/pytorch/pytorch/blob/98fddc1f061fb5b0dcac9f49b3b096bcdc124a12/torch/csrc/autograd/variable.h#L322-L337

I think could add checking `_base` to detect non-leaves to handle that case.

**Comment by albanD** [comment]
Yes views need to inherit the requires grad from their base otherwise code like the following would not work
```python
base = torch.rand(1, 5)
slice1 = base[0]
slice2 = base[0]

# This makes slice1 and base properly require gradients
slice1.copy_(torch.rand(5, requires_grad=True))
# Other views need to also be properly updated by reading the base's status
loss = slice2.sum()
loss.backward()
```


Do I understand properly that the issue is that this view op makes it so that the input to the linear (and thus your cache) is different at each iteration and so a new entry is created every time (and never cleaned up even when the Tensor goes out of scope)?

**Comment by mcarilli** [comment]
> I think checking t._base() to detect non-leaves to handle that case could eliminate 1.

Are you proposing we check `t._base()` somehow in the line where we check for cache eligibility (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/autocast_mode.cpp#L71)?

**Comment by albanD** [comment]
I guess you could check `!t.is_view()` there indeed. If you say that you never want to cache the result for a view of a larger Tensor?
But then what happens if someone actually have weights that are views of a bigger Tensor for some reason? The caching will just miss them completely.
So I don't think you can have something that will work for everything here by just looking at the Tensor. But maybe never caching views will be a good enough heuristic.

**Comment by ezyang** [comment]
Thanks, this is a good analysis. A quick thoughts: view cache blow up due to .t() can be solved with a simple heuristic: if we have a view of a parameter, cast the *source* of the view and cache that. Then, when checking the cache, if we are given a view, check if we have a cache of the view base, and if so, redo the view on top of that.

**Comment by albanD** [comment]
Users might have the bad idea to assign as a parameter a small slice of a much bigger Tensor. Such approach would process a very big Tensor that should not be be processed.
But this would be rare as well.

**Comment by ezyang** [comment]
If this is a concern, you can update the heuristic to only see through view operations that preserve numel.

Also, it is only a problem to cast the bigger tensor *if you don't actually use the rest of the parameters in the bigger tensor in that run*. If you do use them, then you probably are going to autocast them too, in which case you're just paying for it up front.

**Comment by mcarilli** [comment]
> update the heuristic to only see through view operations that preserve numel.

I like this idea, but I don't think it should block https://github.com/pytorch/pytorch/pull/48696, which is a simple fix that's strictly better than current code.

**Comment by stas00** [comment]
May I also ask in this context why caching is skipped if: `!arg.requires_grad()` - e.g. multple transformers models have static positional embedding (not part of the grad tree) which don't ever change so I'd think those should be cached.

And moreover, since they don't ever change these would be a candidate for a constant cache and not re-calculated on each enter/exit from autocast context (if memory allows for that luxury).

Unless there is a way to force tensors to have both float and half values on the user side if the user knows that it'd be beneficial to have those converted values cached once.

I don't know whether there has been a discussion about a session-level cache.

**Comment by mcarilli** [comment]
> May I also ask in this context why caching is skipped if: !arg.requires_grad()

@stas00  `is_leaf()`, as far as I know, checks if the tensor has any autograd history (@albanD knows more).  If we're not under no_grad, you're probably right:  `arg.is_leaf()` suffices to tell us if an incoming param is a static model param in most cases.

But if we are under no_grad, all newly created tensors pass the `arg.is_leaf()` check, because they don't have autograd history.  If we get rid of the `arg.requires_grad()` check, all fp32->fp16 casts for such internal temporaries will be cached, more such temporaries will be cached every iteration of the `forward` loop, and we have another cache blowup.  In this way, under no_grad the `arg.requires_grad()` check is the best way of ensuring we only cache casts for the true static model params, because most** of the temporaries will have `arg.requires_grad() == false` and be ineligible for caching.  It's not a perfect heuristic but I'm not sure how we can make it perfect.  The only information `cached_cast` has is `is_leaf()`, `requires_grad()`, and `is_view()`, and with those it needs to make its best guess whether to cache or not (while hopefully avoiding blowups in calling code).

** the exception is view temporaries like `weight.t()`, which _do_ inherit requires_grad-ness from their parents, so they passed both the is_leaf() and requires_grad() checks to cause the original blowup.

**Comment by stas00** [comment]
Thank you for the detailed explanation, @mcarilli. 

Following this discussion it sounds like the main problem here is that there is an expectation that `autocast` is supposed to magically optimize everything - and everything is different, so as you all have been saying this is an impossible task, and battles have to be picked.

Would the developers consider adding some manual options in addition to the magical `autocast`. For example, could a tensor have a flag that will make it cachable unconditionally?

For example for static positional embeddings we already call `detach` and turn grads off, so we could just as well turn on a hypothetical "cache-these-weights-if-under-fp16" flag. And then another flag for the caching scope `{autocast|session}` - so the tensor will still get cached under `autocast` even if normal conditions for caching won't apply.  And in the case of the `session` scope `autocast` will not release these at the end of its scope.

This is a half-baked idea since I haven't measured the actual impact and the importance of such need. But perhaps some of you who have a lot more practical experience can think of situations where being able to manually control such caching would lead to a faster program at a reasonable memory cost and no memory leaking. I could be wrong and perhaps such need is very minor and is not worthy the added complexity.

So this is not a request but just an idea that I'm asking whether it is reasonable/useful/worthy consideration.

Thank you.

**Comment by albanD** [comment]
Reopening as the PR was just a quick fix and more might need to be done

**Comment by stas00** [comment]
Now that the nightly binary build was updated I verified that both the short reproduction script in the first comment and the original application where I discovered this leak no longer have the leak, and I get about 30% speed increase with `--fp16` - yay!

**Comment by stas00** [comment]
Is this fix going to be added to 1.7.x or just 1.8.0?

**Comment by albanD** [comment]
This is only going to be in 1.8

**Comment by ptrblck** [comment]
Quick side note, it seems the fix landed in `1.7.1` in https://github.com/pytorch/pytorch/pull/48936 🎉

**Comment by stas00** [comment]
That's great news - thank you for the heads up, @ptrblck!

**Comment by mattip** [comment]
Can this be re-closed now that the fix is verified?
